In [3]:
#Librerias
import pandas as pd 
import seaborn as sns 
import numpy as np
import matplotlib.pyplot as plt 

pd.options.display.max_columns = None


In [18]:
CSS = """
.output {
    flex-direction: column;
}
"""
HTML('<style>{}</style>'.format(CSS))

In [ ]:
from IPython.display import display_html
def display_side_by_side(*args):
    html_str=''
    for df in args:
        html_str+=df.to_html()
    display_html(html_str.replace('table','table style="display:inline"'),raw=True)

In [21]:
#Base de datos 
df = pd.read_csv("D:/DATASET/youngpeople.csv")
display(df.shape)
display(df.head(3))


(1010, 150)

Music  Slow songs or fast songs  Dance  Folk  Country  Classical music  \
0    5.0                       3.0    2.0   1.0      2.0              2.0   
1    4.0                       4.0    2.0   1.0      1.0              1.0   
2    5.0                       5.0    2.0   2.0      3.0              4.0   

   Musical  Pop  Rock  Metal or Hardrock  Punk  Hiphop, Rap  Reggae, Ska  \
0      1.0  5.0   5.0                1.0   1.0          1.0          1.0   
1      2.0  3.0   5.0                4.0   4.0          1.0          3.0   
2      5.0  3.0   5.0                3.0   4.0          1.0          4.0   

   Swing, Jazz  Rock n roll  Alternative  Latino  Techno, Trance  Opera  \
0          1.0          3.0          1.0     1.0             1.0    1.0   
1          1.0          4.0          4.0     2.0             1.0    1.0   
2          3.0          5.0          5.0     5.0             1.0    3.0   

   Movies  Horror  Thriller  Comedy  Romantic  Sci-fi  War  \
0     5.0     4.0       2.0     5.0       4.0     4.0  1.0   
1     5.0     2.0       2.0     4.0       3.0     4.0  1.0   
2     5.0     3.0       4.0     4.0       2.0     4.0  2.0   

   Fantasy/Fairy tales  Animated  Documentary  Western  Action  History  \
0                  5.0       5.0          3.0      1.0     2.0      1.0   
1                  3.0       5.0          4.0      1.0     4.0      1.0   
2                  5.0       5.0          2.0      2.0     1.0      1.0   

   Psychology  Politics  Mathematics  Physics  Internet   PC  \
0         5.0       1.0          3.0      3.0       5.0  3.0   
1         3.0       4.0          5.0      2.0       4.0  4.0   
2         2.0       1.0          5.0      2.0       4.0  2.0   

   Economy Management  Biology  Chemistry  Reading  Geography  \
0                 5.0      3.0        3.0      3.0        3.0   
1                 5.0      1.0        1.0      4.0        4.0   
2                 4.0      1.0        1.0      5.0        2.0   

   Foreign languages  Medicine  Law  Cars  Art exhibitions  Religion  \
0                5.0       3.0  1.0   1.0              1.0       1.0   
1                5.0       1.0  2.0   2.0              2.0       1.0   
2                5.0       2.0  3.0   1.0              5.0       5.0   

   Countryside, outdoors  Dancing  Musical instruments  Writing  \
0                    5.0      3.0                  3.0      2.0   
1                    1.0      1.0                  1.0      1.0   
2                    5.0      5.0                  5.0      5.0   

   Passive sport  Active sport  Gardening  Celebrities  Shopping  \
0            1.0           5.0        5.0          1.0       4.0   
1            1.0           1.0        1.0          2.0       3.0   
2            5.0           2.0        1.0          1.0       4.0   

   Science and technology  Theatre  Fun with friends  Adrenaline sports  Pets  \
0                     4.0      2.0               5.0                4.0   4.0   
1                     3.0      2.0               4.0                2.0   5.0   
2                     2.0      5.0               5.0                5.0   5.0   

   Flying  Storm  Darkness  Heights  Spiders  Snakes  Rats  Ageing  \
0     1.0    1.0       1.0      1.0      1.0       5   3.0     1.0   
1     1.0    1.0       1.0      2.0      1.0       1   1.0     3.0   
2     1.0    1.0       1.0      1.0      1.0       1   1.0     1.0   

   Dangerous dogs  Fear of public speaking        Smoking      Alcohol  \
0             3.0                      2.0   never smoked  drink a lot   
1             1.0                      4.0   never smoked  drink a lot   
2             1.0                      2.0  tried smoking  drink a lot   

   Healthy eating  Daily events  Prioritising workload  Writing notes  \
0             4.0           2.0                    2.0            5.0   
1             3.0           3.0                    2.0            4.0   
2             3.0           1.0                    2.0            5

MUSIC PREFERENCES (19) 0:19

MOVIE PREFERENCES (12) 19:31

HOBBIES & INTERESTS (32) 31:63

PHOBIAS (10) 63:73

HEALTH HABITS (3) 73:76

PERSONALITY TRAITS, VIEWS ON LIFE & OPINIONS (57) 76:133

SPENDING HABITS (7) 133:140

DEMOGRAPHICS (10 ) 140:150

# Pre procesado de datos 

### Tipos de variables 

In [1]:
#Variables categóricas (object)
#Ver que tipos de variables hay en el data
print('\033[1m'+"Tipos de variables en el data:")
display(df.dtypes.value_counts()) 


#Ver el numero de niveles de las variables "object"
print('\033[1m'+"Numero de niveles de las variables object: " )
display(df.loc[:, df.dtypes == object].nunique().head(15))


#Mostrar las columnas que son "object" y su numero de niveles 
print ('\033[1m'+"Columnas del data -object- y su numero de niveles: ")
df.loc[:, df.dtypes == object].apply(lambda x: pd.Series(pd.unique(x))).fillna('').head(10)



Tipos de variables en el data:


NameError: name 'df' is not defined

### Valores nulos 

In [143]:
#Valores nulos: 
#Porcentaje de valores nulos de todo el data:
(df.isna().sum().sum())/(len(df)*len(df.columns)*100)

4.013201320132013e-05

In [12]:
from IPython.display import display, HTML
HTML('<style>.output {flex-direction: column;}</style>')

In [184]:
#Porcentaje de valores nulos por columna (Top)
from IPython.display import display, HTML
df_NA_C = df.isnull().sum(axis = 0)*100/ len(df)
df_NA_C = pd.DataFrame(round(df_NA_C.sort_values(ascending=False).head(10),2))
df_NA_C.columns=["Columna_NA_%"]
display(df_NA_C)

#Porcentaje de valores nulos por fila (Top)
df_NA_F = df.isnull().sum(axis = 1)*100/ len(df.columns)
df_NA_F = pd.DataFrame(round(df_NA_F.sort_values(ascending=False).head(10),2))
df_NA_F.columns=["Fila_NA_%"]
display(df_NA_F)

HTML('<style>.output {flex-direction: row;}</style>')

Columna_NA_%
Weight                 1.98
Height                 1.98
Passive sport          1.49
Chemistry              0.99
Geography              0.89
Documentary            0.79
Latino                 0.79
Smoking                0.79
Punk                   0.79
Theatre                0.79

Fila_NA_%
422       5.33
424       4.67
777       4.67
487       4.67
929       4.67
368       4.67
806       4.00
174       4.00
330       4.00
745       4.00

In [120]:
#Porcentaje de valores nulos por fila (Top)
df_NA = df.isnull().sum(axis = 1)*100/ len(df.columns)
round(df_NA.sort_values(ascending=False).head(10),2)

422    5.33
424    4.67
777    4.67
487    4.67
929    4.67
368    4.67
806    4.00
174    4.00
330    4.00
745    4.00
dtype: float64

# Regresión Logística 

In [ ]:
#FUNCIONES
#Aplicar get_dummies a todas las columnas de un data que sean "object"
def dummy_data(df):
    for columna in df.select_dtypes(include=['object']).columns:
        df[columna]  = pd.Categorical(df[columna])
        df_transformed = pd.get_dummies(df)
        return df_transformed
    

In [140]:
#Data
df = pd.read_csv("D:/DATASET/youngpeople.csv")

#Seleccionar variables
X = df.drop("Happiness in life",axis=1)
Y = df["Happiness in life"]
X_columns = X.columns


#Dividir en test/train
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3, random_state = 0)

#Rellenar valores nulos con media de columna 
X_test.fillna(X_test.mean(), inplace=True)

#Aplicar get_dummies a todas las columnas de un data que sean "object"
X_test = dummy_data(X_test)

C:\Anaconda3\envs\Aibloy\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


In [64]:
#Data
df = pd.read_csv("D:/DATASET/youngpeople.csv")
#...............................................................................................
#Aplicar get_dummies a todas las columnas de un data que sean "object"
def dummy_data(df):
    for columna in df.select_dtypes(include=['object']).columns:
        df[columna]  = pd.Categorical(df[columna])
        df_transformed = pd.get_dummies(df)
        return df_transformed
    
df = dummy_data(data)


#...............................................................................................
#Pre procesado de datos 
X = df.drop("Happiness in life",axis=1)
Y = df["Happiness in life"]
X_columns = X.columns

#...............................................................................................
# Rellenar los huecos NA con la media de columna 
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values = np.nan, strategy = 'mean', verbose=0) 
imputer = imputer.fit(X)
X = imputer.transform(X)

#...............................................................................................
#Dividir en test/train
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3, random_state = 0)

X.fillna(df.mean(), inplace=True)
#...............................................................................................

In [68]:
pd.DataFrame(X)

0    1    2    3    4    5    6    7    8    9   10   11   12   13  \
0     5.0  3.0  2.0  1.0  2.0  2.0  1.0  5.0  5.0  1.0  1.0  1.0  1.0  1.0   
1     4.0  4.0  2.0  1.0  1.0  1.0  2.0  3.0  5.0  4.0  4.0  1.0  3.0  1.0   
2     5.0  5.0  2.0  2.0  3.0  4.0  5.0  3.0  5.0  3.0  4.0  1.0  4.0  3.0   
3     5.0  3.0  2.0  1.0  1.0  1.0  1.0  2.0  2.0  1.0  4.0  2.0  2.0  1.0   
4     5.0  3.0  4.0  3.0  2.0  4.0  3.0  5.0  3.0  1.0  2.0  5.0  3.0  2.0   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
1005  5.0  2.0  5.0  2.0  2.0  5.0  4.0  4.0  4.0  3.0  2.0  4.0  1.0  5.0   
1006  4.0  4.0  5.0  1.0  3.0  4.0  1.0  4.0  1.0  1.0  4.0  1.0  1.0  2.0   
1007  4.0  3.0  1.0  1.0  2.0  2.0  2.0  3.0  4.0  1.0  2.0  3.0  3.0  2.0   
1008  5.0  3.0  3.0  3.0  1.0  3.0  1.0  3.0  4.0  1.0  1.0  2.0  2.0  3.0   
1009  5.0  5.0  4.0  3.0  2.0  3.0  3.0  4.0  1.0  1.0  2.0  3.0  1.0  1.0   

       14   15   16   17   18   19   20   21   22   23   24   25   26   27  \
0     3.0  1.0  1.0  1.0  1.0  5.0  4.0  2.0  5.0  4.0  4.0  1.0  5.0  5.0   
1     4.0  4.0  2.0  1.0  1.0  5.0  2.0  2.0  4.0  3.0  4.0  1.0  3.0  5.0   
2     5.0  5.0  5.0  1.0  3.0  5.0  3.0  4.0  4.0  2.0  4.0  2.0  5.0  5.0   
3     2.0  5.0  1.0  2.0  1.0  5.0  4.0  4.0  3.0  3.0  4.0  3.0  1.0  2.0   
4     1.0  2.0  4.0  2.0  2.0  5.0  4.0  4.0  5.0  2.0  3.0  3.0  4.0  4.0   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
1005  4.0  4.0  3.0  3.0  2.0  5.0  2.0  3.0  4.0  5.0  4.0  2.0  5.0  5.0   
1006  3.0  1.0  3.0  4.0  1.0  5.0  2.0  5.0  5.0  1.0  5.0  5.0  1.0  5.0   
1007  2.0  5.0  1.0  1.0  1.0  4.0  5.0  5.0  3.0  5.0  2.0  1.0  4.0  4.0   
1008  2.0  5.0  2.0  2.0  1.0  5.0  3.0  4.0  5.0  3.0  1.0  2.0  5.0  5.0   
1009  3.0  2.0  2.0  3.0  1.0  5.0  2.0  3.0  5.0  2.0  3.0  1.0  4.0  2.0   

       28   29   30   31   32   33   34   35   36   37   38   39   40   41  \
0     3.0  1.0  2.0  1.0  5.0  1.0  3.0  3.0  5.0  3.0  5.0  3.0  3.0  3.0   
1     4.0  1.0  4.0  1.0  3.0  4.0  5.0  2.0  4.0  4.0  5.0  1.0  1.0  4.0   
2     2.0  2.0  1.0  1.0  2.0  1.0  5.0  2.0  4.0  2.0  4.0  1.0  1.0  5.0   
3     5.0  1.0  2.0  4.0  4.0  5.0  4.0  1.0  3.0  1.0  2.0  3.0  3.0  5.0   
4     3.0  1.0  4.0  3.0  2.0  3.0  2.0  2.0  2.0  2.0  2.0  3.0  3.0  5.0   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
1005  2.0  1.0  4.0  4.0  4.0  2.0  1.0  2.0  4.0  3.0  1.0  2.0  1.0  5.0   
1006  5.0  2.0  5.0  4.0  1.0  1.0  5.0  4.0  5.0  5.0  2.0  2.0  1.0  1.0   
1007  3.0  1.0  3.0  2.0  3.0  3.0  3.0  1.0  3.0  4.0  1.0  1.0  1.0  4.0   
1008  3.0  1.0  4.0  3.0  5.0  3.0  1.0  1.0  5.0  4.0  3.0  3.0  3.0  2.0   
1009  4.0  1.0  4.0  2.0  3.0  1.0  2.0  2.0  4.0  2.0  1.0  5.0  3.0  1.0   

       42   43   44   45   46   47   48   49   50   51        52   53   54  \
0     3.0  5.0  3.0  1.0  1.0  1.0  1.0  5.0  3.0  3.0  2.000000  1.0  5.0   
1     4.0  5.0  1.0  2.0  2.0  2.0  1.0  1.0  1.0  1.0  1.000000  1.0  1.0   
2     2.0  5.0  2.0  3.0  1.0  5.0  5.0  5.0  5.0  5.0  5.000000  5.0  2.0   
3     4.0  4.0  2.0  5.0  1.0  5.0  4.0  1.0  1.0  1.0  3.000000  1.0  1.0   
4     2.0  3.0  3.0  2.0  3.0  1.0  4.0  4.0  1.0  3.0  1.000000  3.0  1.0   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...       ...  ...  ...   
1005  4.0  5.0  2.0  1.0  1.0  5.0  1.0  2.0  2.0  3.0  4.000000  4.0  4.0   
1006  2.0  2.0  1.0  1.0  5.0  1.0  5.0  5.0  5.0  5.0  1.901394  1.0  5.0   
1007  1.0  5.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  2.0  1.000000  2.0  2.0   
1008  5.0  5.0  3.0  1.0  1.0  5.0  1.0  5.0  4.0  1.0  4.000000  2.0  4.0   
1009  5.0  4.0  5.0  1.0  3.0  1.0  1.0  4.0  2.0  1.0  2.000000  5.0  4.0   

       55   56   57   58   59   60   61   62   63   64   65   66   67   68  \
0     5.0  1.0  4.0  4.0  2.0  5.0  4.0  4.0  1.0  1.0  1.0  1.0  1.0  5.0   
1     1.0  2.0  3.0  3.0  2.0  4.0  2.0  5.0  1.0  1.0  1.0  2.0  1.0  1.0   
2     1.0  1.0  4.0  2.0  

In [6]:
# Arbol de decisión: Clasificación 
from sklearn.ensemble import RandomForestClassifier
#.......................................................................
random_forest = RandomForestClassifier(
    n_estimators=100                    #n_estimators = Numero de arboles que componen el bosque. Por defecto 10.
)
#.......................................................................
random_forest.fit(X_train, Y_train)

Y_pred = random_forest .predict(X_test)

#Estimado
random_forest.score(X_test, Y_test)

ValueError: could not convert string to float: 'former smoker'